runtime/common/src/mmr.rs

```diff
@@ -0,0 +1,153 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A pallet responsible for creating Merkle Mountain Range (MMR) leaf for current block.
+
+use sp_core::H256;
+use sp_std::prelude::*;
+use frame_support::{decl_error, decl_module, decl_storage, RuntimeDebug,};
+use pallet_mmr::primitives::LeafDataProvider;
+use parity_scale_codec::{Encode, Decode};
+use runtime_parachains::paras;
+
+/// A leaf that get's added every block to the MMR constructed by [pallet_mmr].
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct MmrLeaf<Hash, MerkleRoot> {
+	/// Current block parent hash.
+	pub parent_hash: Hash,
+	/// A merkle root of all registered parachain heads.
+	pub parachain_heads: MerkleRoot,
+	/// A merkle root of current beefy authority set.
+	pub beefy_authority_set: MerkleRoot,
+}
+
+impl<T: Config> LeafDataProvider for Module<T> where
+	MerkleRootOf<T>: From<H256>,
+{
+	type LeafData = MmrLeaf<
+		<T as frame_system::Config>::Hash,
+		MerkleRootOf<T>,
+	>;
+
+	fn leaf_data() -> Self::LeafData {
+		MmrLeaf {
+			parent_hash: frame_system::Module::<T>::leaf_data(),
+			parachain_heads: Module::<T>::parachain_heads_merkle_root(),
+			beefy_authority_set: Module::<T>::beefy_authority_set_merkle_root(),
+		}
+	}
+}
+
+type MerkleRootOf<T> = <T as pallet_mmr::Config>::Hash;
+
+/// The module's configuration trait.
+pub trait Config: pallet_mmr::Config + paras::Config + pallet_beefy::Config
+{}
+
+/// Blanket-impl the trait for every runtime that has both MMR pallet and parachains configuration.
+///
+/// NOTE Remember that you still need to register the [Module] in `construct_runtime` macro.
+impl<R: pallet_mmr::Config + paras::Config + pallet_beefy::Config> Config for R {}
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// No error
+		None,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Beefy {
+		/// The merkle root of next BEEFY authority set.
+		pub BeefyAuthoritiesRoot get(fn beefy_authorities_root): MerkleRootOf<T>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: <T as frame_system::Config>::Origin {
+		type Error = Error<T>;
+	}
+}
+
+impl<T: Config> Module<T> where
+	MerkleRootOf<T>: From<H256>,
+{
+	/// Returns latest root hash of a merkle tree constructed from all registered parachain headers.
+	///
+	/// NOTE this does not include parathreads - only parachains are part of the merkle tree.
+	///
+	/// TODO [ToDr] describe merkle tree construction.
+	///
+	/// NOTE This is an initial and inefficient implementation, which re-constructs
+	/// the merkle tree every block. Instead we should update the merkle root in [Self::on_initialize]
+	/// call of this pallet and update the merkle tree efficiently (use on-chain storage to persist inner nodes).
+	fn parachain_heads_merkle_root() -> MerkleRootOf<T> {
```

Comment:
do we have an rpc or function to get a merkle proof for a particular para_head yet?

Reply:
Not yet, I'm going to evaluate what we can do to expose that.

Reply:
For the record: this has been done as utility library here https://github.com/paritytech/grandpa-bridge-gadget/issues/63